scripts/get_nbd_extents.py

```diff
@@ -0,0 +1,122 @@
+#!/usr/bin/python
+
+"""
+Returns a list of extents with their block statuses for an NBD export.
+
+This program uses new NBD capabilities introduced in QEMU 2.12.
+
+It uses the BLOCK_STATUS NBD extension, which is documented here:
+https://github.com/NetworkBlockDevice/nbd/blob/extension-blockstatus/doc/proto.md
+The structured replies functionality that this extension relies on is in the
+main branch of the NBD protocol docs:
+https://github.com/NetworkBlockDevice/nbd/blob/master/doc/proto.md
+"""
+
+import argparse
+import json
+import logging
+import logging.handlers
+
+from python_nbd_client import PythonNbdClient
+import python_nbd_client
+
+
+LOGGER = logging.getLogger("get_nbd_extents")
+LOGGER.setLevel(logging.DEBUG)
+# The log level of python_nbd_client is not set, therefore it will default to
+# that of the root logger, which is WARNING by default.
+
+# Request length is a 32-bit int.
+# It looks like for qemu 2.12, the length parameter of a NBD_CMD_BLOCK_STATUS
+# request is not limited by the maximum block size supported by the server (as
+# defined by NBD_INFO_BLOCK_SIZE), only by the size of a 32-bit int.
+MAX_REQUEST_LEN = 2 ** 32 - 1
+
+# Make the NBD_CMD_BLOCK_STATUS request aligned to 512 bytes, just in case. But
+# it looks like this is not required for qemu 2.12.
+MAX_REQUEST_LEN = MAX_REQUEST_LEN - (MAX_REQUEST_LEN % 512)
+
+def _get_extents(path, exportname):
+    with PythonNbdClient(
+        address=path, exportname=exportname, unix=True, use_tls=False, connect=False) as client:
+
+        client.negotiate_structured_reply()
+
+        # Select our metadata context. This context is documented at
+        # https://github.com/NetworkBlockDevice/nbd/blob/extension-blockstatus/doc/proto.md#baseallocation-metadata-context
+        context = 'base:allocation'
+        selected_contexts = client.set_meta_contexts(exportname, [context])
+        assert len(selected_contexts) == 1
+        (meta_context_id, meta_context_name) = selected_contexts[0]
+        assert meta_context_name == context
```

Comment:
Note that asserts are removed if python is called with optimizations enabled

Reply:
How can I assert in a way that is always preserved?

Reply:
`if ...: raise AppropriateError`